prow/metrics/metrics.go

```diff
@@ -26,23 +26,51 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/push"
 	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/config"
 )
 
-// PushMetrics is meant to run in a goroutine and continuously push
+type configGetter interface {
+	Config() *config.Config
+}
+
+// Pusher is a struct that knows how to push metrics to prometheus
+type Pusher struct {
+	gatherer    gathererFunc
+	ConfigAgent configGetter
+	waiter      func(time.Duration) <-chan time.Time
+}
+
+// Start is meant to run in a goroutine and continuously push
 // metrics to the provided endpoint.
-func PushMetrics(component, endpoint string, interval time.Duration) {
+func (pm Pusher) Start(component string) {
+	if pm.gatherer == nil {
+		pm.gatherer = push.FromGatherer
+	}
+	if pm.waiter == nil {
+		pm.waiter = time.After
+	}
+
 	sig := make(chan os.Signal, 1)
 	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
 
 	for {
+		config := pm.ConfigAgent.Config()
+		interval := config.PushGateway.Interval
+		waiter := pm.waiter(interval)
+
 		select {
-		case <-time.Tick(interval):
-			if err := push.FromGatherer(component, push.HostnameGroupingKey(), endpoint, prometheus.DefaultGatherer); err != nil {
-				logrus.WithField("component", component).WithError(err).Error("Failed to push metrics.")
+		case <-waiter:
+			endpoint := config.PushGateway.Endpoint
```

Comment:
Reread the config here before using `endpoint`. If `endpoint` changes, you most likely wait in `select` for the timer to kick in with the wrong endpoint so the first push will fail (assuming the old gateway has been removed and only the new endpoint is valid).